### DISCUSSION TITLE
params collections in C#13 and error CS4007

### DISCUSSION THREAD

**Discussion by Aniobodo** [dotnet/csharplang#8600] [Language Ideas]
```
public static IEnumerable<string> EnumerateFiles(string folderPath, params ReadOnlySpan<string> searchPatterns)
{
    foreach (string searchPattern in searchPatterns)
    { foreach (string filePath in Directory.EnumerateFiles(folderPath, searchPattern)) yield return filePath; }
}
```

Error CS4007 is thrown at build time for the method above. The message implies that the instance of the type “System.ReadOnlySpan<string>” cannot be retained beyond the limits “await” or “yield”.	
This is a serious limitation on the params collections feature of C# 13.
So, I wonder if there is a safe way out of the limitation.

**Comment by 333fred** [comment]
Not if you want to use a generated iterator. You need either evaluate everything up front, and return all the results at once, or pay for the search patterns alloe. Imo, you're already paying for the iterator allocation, so the additional memory for the patterns really isn't much more.

**Reply by colejohnson66** [reply]
To elaborate: it's a memory-safety issue. Consider this code:

```csharp
public void DoStuff()
{
    IEnumerable<string> files = GetFiles();
    // process 'files'
}
public IEnumerable<string> GetFiles()
{
    return EnumerateFiles("<folder>", "<search-pattern-1>", "<search-pattern-2>", "<search-pattern-3>");
}
```

The implicitly created `ReadOnlySpan<string>` in `GetFiles` is only valid *inside* `GetFiles`. Returning it further up the stack into `DoStuff` would cause the span to be pointing at garbage on the stack, which is a memory-safety no-no. It's the same problem as this:

```csharp
public ref int GetInt()
{
    int i = 42;
    return ref i;
}
```

**Comment by timcassell** [comment]
You can use `ArrayPool` to rent an array for use in the iterator. Though as @333fred mentioned, you're already allocating for the iterator, so it's maybe not worth it, as you can just allocate an array.
